src/util/cilfacade.ml

```diff
@@ -79,6 +79,81 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+class loopUnrollingVisitor = object
+  inherit nopCilVisitor
+  method! vstmt s =
+    (* All the duplicated stmts need to be re-created or the sid will fail*)
+    let newsid st = { st with sid = st.sid } in
+    let rec newsid_stmt st = 
+      let mkb stml = mkBlock (List.map newsid_stmt stml) in
+      match st.skind with
+      | Instr _ -> newsid st
+      | If(e,b1,b2,l1,l2) -> mkStmt(If(e,(mkb b1.bstmts),(mkb b2.bstmts),l1,l2))
+      | Loop(bl,l1,l2,s1,s2) -> 
+        let create_stmt = mkStmt(Loop((mkb bl.bstmts),l1,l2,s1,s2)) in
+        create_stmt.labels <- st.labels;
+        create_stmt
+      | Block(bl) -> mkStmt(Block(mkb bl.bstmts))
+      | Switch(e,bl,stl,l1,l2) -> mkStmt(Switch(e,(mkb bl.bstmts),(List.map newsid_stmt stl),l1,l2))
+      | _ -> newsid st in
+    let mkBlock' b = mkBlock (List.map newsid_stmt b) in
+    let mk_stmt_from_stmt_list bl = mkStmt(Block(mkBlock' bl)) in
```

Comment:
This looks like part of `copyFunctionVisitor` in CIL, which seems to handle `Goto`s for the copies as well. Maybe something could be extracted from there.

Reply:
I took a look at it and I don't see a way to use the `copyFunctionVisitor` directly. The only option I see is replicating what happens there, which is what I did. Do you maybe have a suggestion on how to approach this?

Reply:
It probably cannot be used directly, because it's meant for copying entire functions and that's how it's also exposed from CIL. But I was thinking something along the lines of copying its code from CIL here (as `copyStmtVisitor` or something) and simplifying it a bit (removing the function-specific parts and maybe removing `Switch` support, which are handled before by `prepareCFG` anyway).

The nice thing about such visitor is that it doesn't need to handle all `stmtkind`s separately and recurse into them, but the visitor logic itself already takes care of all of that. And the `patches` component of it could also be kept to fix `Goto`s for the copies. I think that will inevitably be needed because the prepared switches and loops contain them. I haven't tried, but I feel like currently things might go wrong if the unrolled loop contains a switch or another loop.